src/Data/Conduit/Tar.hs

```diff
@@ -270,3 +245,482 @@ withEntries :: MonadThrow m
+                when (headerMagicVersion h == gnuTarMagicVersion) $ handleGnuTarHeader h
+                withFileInfo inner
+        Just (ChunkHeader h) -> do
+            payloadsConduit .| (inner (fileInfoFromHeader h) <* sinkNull)
+            withFileInfo inner
+        Just x@(ChunkPayload offset _bs) -> do
+            leftover x
+            throwM $ UnexpectedPayload offset
+        Just (ChunkException e) -> throwM e
+
+
+-- | Take care of custom GNU tar format.
+handleGnuTarHeader :: MonadThrow m
+                   => Header
+                   -> ConduitM TarChunk o m ()
+handleGnuTarHeader h = do
+    case headerLinkIndicator h of
+        76 -> do
+            let pSize = headerPayloadSize h
+            -- guard against names that are too long in order to prevent a DoS attack on unbounded
+            -- file names
+            unless (0 < pSize && pSize <= 4096) $
+                throwM $
+                FileTypeError (headerPayloadOffset h) 'L' $ "Filepath is too long: " ++ show pSize
+            longFileNameBuilder <- payloadsConduit .| sinkBuilder
+            let longFileName = SL.toStrict . SL.init . toLazyByteString $ longFileNameBuilder
+            mcNext <- await
+            case mcNext of
+                Just (ChunkHeader nh) -> do
+                    unless (S.isPrefixOf (fromShort (headerFileNameSuffix nh)) longFileName) $
+                        throwM $
+                        FileTypeError (headerPayloadOffset nh) 'L' $
+                        "Long filename doesn't match the original."
+                    leftover
+                        (ChunkHeader $
+                         nh
+                         { headerFileNameSuffix = toShort longFileName
+                         , headerFileNamePrefix = SS.empty
+                         })
+                Just c@(ChunkPayload offset _) -> do
+                    leftover c
+                    throwM $ InvalidHeader offset
+                Just (ChunkException exc) -> throwM exc
+                Nothing -> throwM NoMoreHeaders
+        83 -> do
+            payloadsConduit .| sinkNull -- discard sparse files payload
+            -- TODO : Implement restoring of sparse files
+        _ -> return ()
+
+
+-- | Just like `withFileInfo`, but works directly on the stream of bytes.
+untar :: MonadThrow m
+      => (FileInfo -> ConduitM ByteString o m ())
+      -> ConduitM ByteString o m ()
+untar inner = untarChunks .| withFileInfo inner
+
+
+-- | Just like `untar`, except that each `FileInfo` handling function can produce a finalizing
+-- action, all of which will be executed after the whole tarball has been processed in the opposite
+-- order. Very useful with `restoreFile` and `restoreFileInto`, since they restore direcory
+-- modification timestamps only after files have been fully written to disk.
+untarWithFinalizers ::
+       (MonadThrow m, MonadIO m)
+    => (FileInfo -> ConduitM ByteString (IO ()) m ())
```

Comment:
Interesting design choice to put the finalizers into the output stream instead of the result values. If only some of the actions are intended to return finalizers, that can be a win by avoiding holding onto `return ()` actions. Is that the situation here?

Reply:
Yes, this is the case. Finalizers are generated only for directories, not for regular files or symlinks.
https://github.com/snoyberg/tar-conduit/pull/10/files#diff-6f0cb0cfde65716bd733c518156d3769R62